client/client.go

```diff
@@ -44,54 +45,52 @@ type Client struct {
 // It establishes a connection to the blockchain and verifies the integrity of contracts at the given address.
 // It uses the comm backend to initialize adapters for off-chain communication network.
 func NewEthereumPaymentClient(cfg Config, user perun.User, comm perun.CommBackend) (*Client, error) {
-	dialer, listener, err := initComm(comm, user.CommAddr)
-	if err != nil {
-		return nil, err
-	}
 	funder, adjudicator, err := connectToChain(cfg.Chain, user.OnChain)
 	if err != nil {
 		return nil, err
 	}
-	// Only off-chain account is unlocked. Accounts for the participant addresses
-	// will be unlocked by the client when required.
 	offChainAcc, err := user.OffChain.Wallet.Unlock(user.OffChain.Addr)
 	if err != nil {
 		return nil, errors.WithMessage(err, "off-chain account")
 	}
+	msgBus := net.NewBus(offChainAcc, comm.NewDialer())
 
-	c := client.New(offChainAcc, dialer, funder, adjudicator, user.OffChain.Wallet)
-	if err := loadPersister(c, cfg.DatabaseDir, cfg.PeerReconnTimeout); err != nil {
+	c, err := client.New(offChainAcc.Address(), msgBus, funder, adjudicator, user.OffChain.Wallet)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing state channel client")
+	}
+	if err = loadPersister(c, cfg.DatabaseDir, cfg.PeerReconnTimeout); err != nil {
 		return nil, err
 	}
+
 	client := &Client{
 		Client: c,
+		msgBus: msgBus,
 		wg:     &sync.WaitGroup{},
 	}
 
-	client.runAsGoRoutine(func() { client.Handle(&proposalHandler{}, &updateHandler{}) })
-	client.runAsGoRoutine(func() { client.Listen(listener) })
+	listener, err := comm.NewListener(user.CommAddr)
+	if err != nil {
+		return nil, err
+	}
+	client.runAsGoRoutine(func() { client.Handle(&ProposalHandler{}, &UpdateHandler{}) })
+	client.runAsGoRoutine(func() { msgBus.Listen(listener) })
 
 	return client, nil
 }
 
 // Close closes the client and waits until the listener and handler go routines return.
 //
-// Close depends on the following mechanisms implemented in client.Close:
-// 1. A context is passed to the infinite loop of handle function. When client.Close is called,
-//    this context is canceled, thereby causing the infinite loop in Handle function to stop.
-// 2. A callback to close the listener (listener.Close()) is registered to the client.Closer.
-//    When client.Close is called, this call back signals the listener to shutdown and hence
-//    stop the infinite loop in Listen function.
+// Close depends on the following mechanisms implemented in client.Close and bus.Close to signal the go-routines:
+// 1. When client.Close is invoked, it signals the infinite loop (via a context) in Handle function to stop.
```

Comment:
```suggestion
// 1. When client.Close is invoked, it cancels the Update and Proposal handlers via a context.
```

Reply:
Updated, in cc40df19408154579443ce9af38e9aa743e8c9a2.